File: src/_plugins/social_embeds.rb
# This is a plugin for embedding social media posts that avoids using
# native embeds. Rendering posts as static HTML reduces page weight,
# load times, and is resilient against posts being deleted.
#
# To embed a post, use a Liquid tag of the form:
#
#     {% mastodon https://code4lib.social/@linguistory/113924700205617006 %}
#

require_relative 'utils/twitter'

module Jekyll
  module SocialEmbedFilters
    # Create a data URI for the avatar.
    #
    # These images are tiny when resized properly – in most cases <4KB,
    # so it's faster to embed them as base64-encoded images than serve
    # them as a separate network request.
    def avatar_url(post_data)
      # Avatars are stored in the _images/social_embeds/avatars directory,
      # combining the site, user ID, and post ID.
      case post_data['site']
      when 'mastodon'
        user_id = post_data['author']['username']
        post_id = post_data['id']
      when 'twitter'
        user_id = post_data['author']['screen_name']
        post_id = post_data['id']
      when 'bluesky'
        user_id = post_data['author']['handle']
        post_id = post_data['id']
      else
        raise "Unrecognised site: #{post_data['site']}"
      end

      avatar_id = "#{user_id}_#{post_id}"

      Jekyll::Cache.new('SocialEmbedAvatars').getset(avatar_id) do
        # Find the matching avatar, which may be one of several formats.
        matching_avatars = Dir.glob("src/_images/social_embeds/avatars/#{avatar_id}*")

        unless matching_avatars.length == 1
          raise "Could not find avatar for #{avatar_id}"
        end

        avatar_path = matching_avatars[0]
        create_base64_avatar(avatar_path, 92)
      end
    end

    # render_mastodon_text renders the text of a Mastodon post as HTML.
    #
    # This includes:
    #
    #   - Expanding newlines
    #   - Adding hashtags
    #
    def render_mastodon_text(post_data)
      # Newlines aren't significant in HTML; convert them to <br> tags.
      text = post_data['text']
      text = text.gsub("\n", '<br>')

      # Replace any hashtags with links to the server.
      server = post_data['author']['server']
      entities = post_data.fetch('entities', {})
      entities.fetch('hashtags', []).each do |h|
        text = text.sub(
          "##{h}",
          "<a href=\"https://#{server}/tags/#{h}\">##{h}</a>"
        )
      end

      # Replace any URLs with their display versions.
      entities.fetch('urls', []).each do |u|
        text = text.sub(
          u['url'],
          "<a href=\"#{u['url']}\">#{u['display_url']}</a>"
        )
      end

      # Replace any user mentions with links to the username.
      entities.fetch('user_mentions', []).each do |u|
        text = text.sub(
          "@#{u['label']}",
          "<a href=\"#{u['profile_url']}\">@#{u['label']}</a>"
        )
      end

      text.strip
    end

    # tweet_image creates a {% picture %} tag to show a media item
    # on a tweet.
    def tweet_image(media)
      alt_text = media['ext_alt_text']

      <<~HTML
        <a href="#{media['url']}">
          {%
            picture
            filename="#{media['filename']}"
            parent="/images/social_embeds/twitter"
            #{alt_text.nil? ? 'data-proofer-ignore' : "alt=\"#{alt_text}\""}
            width="585"
          %}
        </a>
      HTML
    end

    # render_tweet_text renders the text of a tweet as HTML.
    #
    # This includes:
    #
    #     * Expanding any newlines
    #     * Replacing URLs and @-mentions
    #     * Replacing native emoji with Twitter's "twemoji" SVGs
    #
    def render_tweet_text(post_data)
      text = post_data['text']

      entities = post_data.fetch('entities', {})

      # Expand any t.co URLs in the text with the actual link, which means
      # those links don't rely on Twitter or their link shortener.
      entities.fetch('urls', []).each do |u|
        text = text.sub(
          u['url'],
          "<a href=\"#{u['url']}\">#{u['display_url']}</a>"
        )
      end

      # Because newlines aren't significant in HTML, we convert them to
      # <br> tags so they render correctly.
      text = text.gsub("\n", '<br>')

      # Ensure user mentions (e.g. @alexwlchan) in the body of the tweet
      # are correctly rendered as links to the user page.
      entities.fetch('user_mentions', []).each do |um|
        text = text.sub(
          "@#{um}",
          "<a href=\"https://twitter.com/#{um}\">@#{um}</a>"
        )
      end

      entities.fetch('hashtags', []).each do |h|
        text = text.sub(
          "##{h}",
          "<a href=\"https://twitter.com/hashtag/#{h}\">##{h}</a>"
        )
      end

      text.strip
    end

    def replace_twemoji(text)
      replace_twemoji_with_images(text)
    end
  end
end

def create_base64_avatar(path, size)
  require 'base64'
  require 'vips'

  im = Vips::Image.new_from_file path

  if im.width != im.height
    raise "Avatar is not square: #{path}"
  end

  # Resize the image to match the target size
  scale = size.to_f / im.width
  resized = im.resize(scale)

  # Now write the image to a buffer, and convert it to base64.
  #
  # We preserve the original format, which is likely to be the most
  # efficient encoding for this image.
  case im.get 'vips-loader'
  when 'jpegload'
    jpeg_bytes = resized.write_to_buffer('.jpg')
    base64_string = Base64.strict_encode64(jpeg_bytes)
    "data:image/jpeg;base64,#{base64_string}"
  when 'pngload'
    png_bytes = resized.write_to_buffer('.png')
    base64_string = Base64.strict_encode64(png_bytes)
    "data:image/png;base64,#{base64_string}"
  else
    raise "Unrecognised vips loader for #{path}: #{im.get 'vips-loader'}"
  end
end

class SocialMedia < Liquid::Tag
  def initialize(_tag_name, text, _tokens)
    # The `text` variable should contain the URL of the post.
    @post_url = text.strip.delete_prefix('"').delete_suffix('"')
  end

  def render(context)
    # Look up the data for this post in `src/_data/social_embeds.json`,
    # which is keyed by the URL.
    site = context.registers[:site]
    post_data = site.data['social_embeds'][@post_url]

    if post_data.nil?
      raise "Could not find data for URL: #{@post_url}"
    end

    tpl = Liquid::Template.parse(
      File.read("src/_includes/embeds/#{post_data['site']}.html")
    )

    input = tpl.render!(
      'post_url' => @post_url,
      'post_data' => post_data
    )

    # Run it through the site's Markdown converter after rendering
    # the initial HTML, so we have access to the picture plugin.
    Liquid::Template.parse(input).render!(context)
  end
end

Liquid::Template.register_filter(Jekyll::SocialEmbedFilters)
Liquid::Template.register_tag('bluesky', SocialMedia)
Liquid::Template.register_tag('mastodon', SocialMedia)
Liquid::Template.register_tag('tweet', SocialMedia)
Liquid::Template.register_tag('youtube', SocialMedia)

File: mosaic/templates/social_embeds.py
"""
Embed posts from social media websites in my blog.

These embeds are lightweight HTML and CSS rather than native embeds,
so they don't pull in gross JavaScript tracking and are more resilient
to site outages or link rot.

To embed a post, use a tag of the form:

    {% mastodon https://code4lib.social/@linguistory/113924700205617006 %}

== How it works ==

I have a file `social_embeds.json` with data about each embedded post,
which is keyed by URL and filled in by hand.

When I use the {% social %} tag, the data from that file is combined
with a template to render the post.
"""

import base64
import glob
from io import BytesIO
import json
from pathlib import Path

from jinja2 import nodes, pass_context
from jinja2.ext import Extension
from jinja2.parser import Parser
from jinja2.runtime import Context
from PIL import Image

from mosaic.text import assert_is_invariant_under_markdown

from .pictures import render_picture
from .social_embed_models import (
    MastodonEmbed,
    MediaEntity,
    SocialEmbedData,
    TwitterEmbed,
    parse_social_embed_data,
)


with open("src/_data/social_embeds.json") as in_file:
    SOCIAL_EMBEDS_DATA = {
        url: parse_social_embed_data(data) for url, data in json.load(in_file).items()
    }


class SocialExtension(Extension):
    """
    Defines the {% social %} extension.
    """

    tags = {"bluesky", "mastodon", "tweet", "youtube"}

    def parse(self, parser: Parser) -> nodes.Node:
        """
        Parse a single URL from the tag.
        """
        first_token = next(parser.stream)
        assert first_token.value in self.tags, f"first token = {first_token!r}"

        lineno = first_token.lineno

        # Capture the next token as a string (the URL)
        url_token = next(parser.stream)
        url = nodes.Const(url_token.value)

        # Call the helper method to turn the URL into HTML
        call = self.call_method("_render_social", [url], lineno=lineno)

        # Return an Output node so it prints to the page
        return nodes.Output([call], lineno=lineno)

    @pass_context
    def _render_social(self, context: Context, post_url: str) -> str:
        """
        Render the social embed.
        """
        post_data = SOCIAL_EMBEDS_DATA[post_url]
        env = context.environment

        env.filters.update(
            {
                "avatar_url": avatar_url,
                "render_bluesky_text": render_bluesky_text,
                "render_mastodon_text": render_mastodon_text,
                "render_tweet_text": render_tweet_text,
                "replace_twemoji": replace_twemoji,
                "tweet_image": tweet_image,
            }
        )

        template = env.get_template(f"embeds/{post_data.site}.html")
        html = template.render(post_url=post_url, post_data=post_data)

        assert_is_invariant_under_markdown(html)
        return html


@pass_context
def tweet_image(context: Context, media: MediaEntity) -> str:
    """
    Render a media item on a tweet.
    """
    html = render_picture(
        context,
        filename=media.filename,
        parent="/images/social_embeds/twitter",
        alt=media.ext_alt_text or "",
        width=585,
        link_to=media.url,
    )

    html = html.replace('alt=""', "data-proofer-ignore")

    return html


def avatar_url(post_data: SocialEmbedData) -> str:
    """
    Create a data URI for an avatar.

    These images are tiny when resized properly – in most cases <4KB,
    so it's faster to embed them as base64-encoded images than serve
    them as a separate network request.
    """
    match post_data.site:
        case "mastodon":
            user_id = post_data.author.username
            post_id = post_data.id
        case "twitter":
            user_id = post_data.author.screen_name
            post_id = post_data.id
        case "bluesky":
            user_id = post_data.author.handle
            post_id = post_data.id
        case _:  # pragma: no cover
            raise ValueError(f"Unrecognised site: {post_data.site}")

    avatar_id = f"{user_id}_{post_id}"

    # TODO: Implement caching
    # TODO: Allow choosing the source directory here
    matching_avatars = glob.glob(f"src/_images/social_embeds/avatars/{avatar_id}*")

    if len(matching_avatars) != 1:  # pragma: no cover
        raise RuntimeError(f"could not find avatar for {avatar_id}")

    avatar_path = Path(matching_avatars[0])
    return create_base64_avatar(avatar_path, size=92)


def create_base64_avatar(avatar_path: Path, *, size: int) -> str:
    """
    Converts a square avatar to a base64-encoded data URI at the given size.
    """
    with Image.open(avatar_path) as im:
        assert im.width == im.height
        assert im.width >= size

        resized = im.resize((size, size))

        buffer = BytesIO()
        resized.save(buffer, format=im.format)

    base64_string = base64.b64encode(buffer.getvalue()).decode("utf8")

    match im.format:
        case "JPEG":
            return f"data:image/jpeg;base64,{base64_string}"
        case "PNG":
            return f"data:image/png;base64,{base64_string}"
        case _:  # pragma: no cover
            raise RuntimeError(f"unsupported avatar format: {im.format}")


def render_bluesky_text(post_data: TwitterEmbed) -> str:
    """
    Render the text of a Bluesky post as HTML.
    """
    text = post_data.text

    # Newlines aren't significant in HTML; convert them to <br> tags
    # so they render correctly.
    text = text.replace("\n", "<br>")

    return text


def render_mastodon_text(post_data: MastodonEmbed) -> str:
    """
    Render the text of a Mastodon post as HTML.
    """
    text = post_data.text

    # Newlines aren't significant in HTML; convert them to <br> tags
    # so they render correctly.
    text = text.replace("\n", "<br>")

    if post_data.entities is None:
        return text

    server = post_data.author.server

    for h in post_data.entities.hashtags:
        text = text.replace(f"#{h}", f'<a href="https://{server}/tags/{h}">#{h}</a>')

    for um in post_data.entities.user_mentions:
        text = text.replace(
            f"@{um.label}", f'<a href="{um.profile_url}">@{um.label}</a>'
        )

    for u in post_data.entities.urls:
        text = text.replace(u.url, f'<a href="{u.url}">{u.display_url}</a>')

    return text


def render_tweet_text(post_data: TwitterEmbed) -> str:
    """
    Render the text of a tweet as HTML.
    """
    text = post_data.text

    # Newlines aren't significant in HTML; convert them to <br> tags
    # so they render correctly.
    text = text.replace("\n", "<br>")

    if post_data.entities is None:
        return text

    for h in post_data.entities.hashtags:
        text = text.replace(
            f"#{h}", f'<a href="https://twitter.com/hashtag/{h}">#{h}</a>'
        )

    for um in post_data.entities.user_mentions:
        text = text.replace(f"@{um}", f'<a href="https://twitter.com/{um}">@{um}</a>')

    for u in post_data.entities.urls:
        text = text.replace(u.url, f'<a href="{u.url}">{u.display_url}</a>')

    return text


def replace_twemoji(text: str) -> str:
    """
    Replace emoji in tweets with their "twemoji" counterparts.
    """
    # Rather than record the entire twemoji set here, I just have a
    # hard-coded set of rules for the twemoji I know I'm using.
    twemoji = {
        "😎": "1f60e.svg",
        "👌": "1f44c.svg",
        "🐦": "1f426.svg",
        "💻": "1f4bb.svg",
        "🥳": "1f973.svg",
        # NOTE: this is a surfer emoji with skin tone/gender modifiers
        "🏄🏻‍♂️": "1f3c4-1f3fb-200d-2642-fe0f.svg",
        "📈": "1f4c8.svg",
        "💞": "1f49e.svg",
        "🧵": "1f9f5.svg",
        "✨": "2728.svg",
    }

    for orig, svg_name in twemoji.items():
        if orig not in text:
            continue

        # TODO: Allow choosing the source directory
        svg_path = Path("src/_images/social_embeds/twemoji") / svg_name
        base64_string = base64.b64encode(svg_path.read_bytes()).decode("ascii")
        data_uri = f"data:image/svg+xml;base64,{base64_string}"

        # Construct the <img> tag.  Notes:
        #
        #   - the `twemoji` class is used for styling
        #   - the `width`/`height` attributes are meant for when styles
        #     don't load properly -- they stop the emoji completely blowing
        #     up in size.
        #
        text = text.replace(
            orig,
            f'<img class="twemoji" width="20px" height="20px" alt="{orig}" '
            f'src="{data_uri}"/>',
        )

    return text
